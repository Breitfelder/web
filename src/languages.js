export default {
  python: {
    name: 'Python',
    url: 'https://github.com/bblfsh/python-driver',
    mode: 'python'
  },
  java: {
    name: 'Java',
    url: 'https://github.com/bblfsh/java-driver',
    mode: 'text/x-java'
  }
};
