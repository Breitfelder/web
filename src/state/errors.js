export const initialState = [];

export const ADD = 'bblfsh/errors/ADD';
export const SET = 'bblfsh/errors/SET';
export const REMOVE = 'bblfsh/errors/REMOVE';
export const CLEAR = 'bblfsh/errors/CLEAR';

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD:
      return state.concat(action.errors || []);
    case SET:
      return action.errors || [];
    case REMOVE:
      return [...state.slice(0, action.idx), ...state.slice(action.idx + 1)];
    case CLEAR:
      if (action.errorType) {
        return state.filter(e => e.type !== action.errorType);
      }
      return [];
    default:
      return state;
  }
};

function error(obj, type) {
  return {
    type,
    message: obj.toString(),
  };
}

export const add = (errors, type) => ({
  type: ADD,
  errors: type ? errors.map(e => error(e, type)) : errors,
});

export const set = (errors, type) => ({
  type: SET,
  errors: type ? errors.map(e => error(e, type)) : errors,
});

export const remove = idx => ({
  type: REMOVE,
  idx,
});

export const clear = type => ({
  type: CLEAR,
  errorType: type,
});
