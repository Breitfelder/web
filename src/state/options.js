export const initialState = {
  showLocations: false,
  customServer: false,
  customServerUrl: '',
  uastQuery: '',
};

export const LOCATIONS_TOGGLE = 'bblfsh/options/LOCATIONS_TOGGLE';
export const CUSTOM_SERVER_URL_SET = 'bblfsh/options/CUSTOM_SERVER_URL_SET';
export const CUSTOM_SERVER_TOGGLE = 'bblfsh/options/CUSTOM_SERVER_TOGGLE';
export const SET_UAST_QUERY = 'bblfsh/options/SET_UAST_QUERY';

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case LOCATIONS_TOGGLE:
      return {
        ...state,
        showLocations: !state.showLocations,
      };
    case CUSTOM_SERVER_URL_SET:
      return {
        ...state,
        customServerUrl: action.url,
      };
    case CUSTOM_SERVER_TOGGLE:
      return {
        ...state,
        customServer: !state.customServer,
        customServerUrl: state.customServer ? '' : '0.0.0.0:9432',
      };
    case SET_UAST_QUERY:
      return {
        ...state,
        uastQuery: action.query,
      };
    default:
      return state;
  }
};

export const locationsToggle = () => ({ type: LOCATIONS_TOGGLE });

export const setCustomServerUrl = url => ({
  type: CUSTOM_SERVER_URL_SET,
  url,
});
export const customServerToggle = () => ({ type: CUSTOM_SERVER_TOGGLE });
export const setUastQuery = query => ({
  type: SET_UAST_QUERY,
  query,
});

export const isUrl = url => /^[a-zA-Z0-9][^/]+$/.test(url);
