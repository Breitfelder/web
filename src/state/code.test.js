import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

import {
  reducer,
  initialState,
  SET,
  PARSE,
  PARSE_FAILED,
  SET_AST,
  CHANGE,
  runParserWithQuery,
  runParser,
  setAst,
} from './code';
import { set as languageSet } from './languages';
import { setUastQuery } from './options';
import { clear as errorsClear } from './errors';

describe('code/reducer', () => {
  it('SET', () => {
    expect(
      reducer(initialState, {
        type: SET,
        filename: 'file.py',
        code: 'foo = 1',
        ast: { InternalType: 'file' },
      })
    ).toMatchSnapshot();
  });

  it('PARSE', () => {
    expect(
      reducer(
        {
          ...initialState,
          ast: { 1: { id: 1 } },
          posIndex: 'index',
          parsing: false,
        },
        {
          type: PARSE,
        }
      )
    ).toMatchSnapshot();
  });

  it('PARSE_FAILED', () => {
    expect(
      reducer({ ...initialState, parsing: true }, { type: PARSE_FAILED })
    ).toMatchSnapshot();
  });

  it('SET_AST', () => {
    expect(
      reducer(initialState, { type: SET_AST, ast: { 1: { id: 1 } } })
    ).toMatchSnapshot();
  });

  it('CHANGE', () => {
    expect(
      reducer(
        { ...initialState, code: 'foo = 1' },
        { type: CHANGE, code: 'bar = 1' }
      )
    ).toMatchSnapshot();
  });
});

describe('code/actions', () => {
  it('runParserWithQuery', () => {
    const uast = { InternalType: 'root' };
    fetch.mockResponse(JSON.stringify({ status: 0, uast, language: 'python' }));

    const store = mockStore({
      code: { code: 'foo = 1', filename: 'test.py', query: 'test-query' },
      options: { mode: 'semantic', customServer: false, customServerUrl: '' },
      languages: { selected: '' },
      versions: { loadedFrom: undefined },
    });
    const expectedActions = [
      errorsClear('parse'),
      { type: PARSE },
      languageSet('python'),
      setAst(uast),
    ];

    return store.dispatch(runParserWithQuery()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('runParser', () => {
    const uast = { InternalType: 'root' };
    fetch.mockResponse(JSON.stringify({ status: 0, uast, language: 'python' }));

    const store = mockStore({
      code: { code: 'foo = 1', filename: 'test.py' },
      options: { mode: 'semantic', customServer: false, customServerUrl: '' },
      languages: { selected: '' },
      versions: { loadedFrom: undefined },
    });
    const expectedActions = [
      setUastQuery(''),
      errorsClear('parse'),
      { type: PARSE },
      languageSet('python'),
      setAst(uast),
    ];

    return store.dispatch(runParser()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
