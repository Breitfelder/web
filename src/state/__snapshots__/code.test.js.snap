// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code/reducer CHANGE 1`] = `
Object {
  "code": "bar = 1",
  "dirty": true,
  "filename": undefined,
  "parsing": false,
  "uastJson": null,
}
`;

exports[`code/reducer PARSE 1`] = `
Object {
  "code": null,
  "dirty": false,
  "filename": undefined,
  "parsing": true,
  "posIndex": null,
  "uastJson": null,
}
`;

exports[`code/reducer PARSE_FAILED 1`] = `
Object {
  "code": null,
  "dirty": false,
  "filename": undefined,
  "parsing": false,
  "uastJson": null,
}
`;

exports[`code/reducer SET 1`] = `
Object {
  "code": "foo = 1",
  "dirty": false,
  "filename": "file.py",
  "markRange": null,
  "parsing": false,
  "uastJson": Object {
    "InternalType": "file",
  },
}
`;

exports[`code/reducer SET_AST 1`] = `
Object {
  "code": null,
  "dirty": false,
  "filename": undefined,
  "parsing": false,
  "uastJson": Object {
    "1": Object {
      "id": 1,
    },
  },
}
`;
